Add LogixNG action ActionLayoutTurnout

danielb987 opened the pull request:
This PR adds the action LayoutTurnout to the `Display` category. It lets the user to enable/disable a turnout on a layout panel.

It solves this user request:
https://groups.io/g/jmriusers/message/211107

danielb987 commented:
@dsand47 
This PR is ready for review. It adds an action to LogixNG that enables/disables a turnout on the layout editor.

dsand47 commented:
I think it would make sense to select the LE turnout item by the turnout table bean instead of the layout editor id.  Turnout items on the panel without an assigned turnout table bean don't need to be disabled since they are not attached to anything.

The action is designed to affect a single LE turnout item.  This works fine for individual turnout items, but creates a lot of work when the goal is to enable/disable all turnout items on a panel.

By using the bean instead of the `id`, it is possible to use the `for each` turnout action to change all of the turnouts on the panel.  Since a turnout bean might not be used on the selected panel, that is not an error and can be skipped.

dsand47 commented:
The combo box is defaulting to 8 rows.

Using a local variable for the turnout results in `layoutTurnout is null`.  At least in my approach, the local variable has a turnout bean.  The needs to be used to find the `layoutTurnout` for he panel, if any.

danielb987 commented:
> The combo box is defaulting to 8 rows.

Fixed for both panel combo box and item combo box in both Positionable and LayoutTurnout actions.


> Using a local variable for the turnout results in `layoutTurnout is null`. At least in my approach, the local variable has a turnout bean. The needs to be used to find the `layoutTurnout` for he panel, if any.

Fixed.

dsand47 commented on java/src/jmri/jmrit/display/logixng/ActionLayoutTurnout.java:
```diff
@@ -0,0 +1,499 @@
+        } else {
+            _layoutEditor = null;
+        }
+//        InstanceManager.turnoutManagerInstance().addVetoableChangeListener(this);
```
Is it a problem if the layout editor panel has been deleted?

danielb987 replied:
I have added a listener that listens to `EditorManager` and clears the reference to the editor and layout turnout if the editor is deleted.

dsand47 commented on java/src/jmri/jmrit/display/logixng/ActionLayoutTurnout.java:
```diff
@@ -0,0 +1,499 @@
+    }
+
+    public LayoutTurnout findLayoutTurnout(String name) {
+        for (LayoutTurnout lt : _layoutEditor.getLayoutTurnouts()) {
```
_layoutEditor might be null.

danielb987 replied:
Fixed.

danielb987 commented on java/src/jmri/jmrit/display/logixng/ActionLayoutTurnout.java:
```diff
@@ -0,0 +1,499 @@
+    }
+
+    public LayoutTurnout findLayoutTurnout(jmri.Turnout turnout) {
+        for (LayoutTurnout lt : _layoutEditor.getLayoutTurnouts()) {
```
Fixed.

danielb987 commented on java/src/jmri/jmrit/display/logixng/ActionLayoutTurnout.java:
```diff
@@ -0,0 +1,499 @@
+//        System.out.format("ActionLayoutTurnout.execute: layoutTurnout: %s%n", layoutTurnout);
+
+        if (layoutTurnout == null) {
+            log.error("layoutTurnout is null");
```
Fixed.

dsand47 reviewed (approved):
This will be useful, especially with `for each`, when switching between local control with active panels and dispatcher control with panel actions disabled.

danielb987 commented:
@dsand47 
> This will be useful, especially with `for each`, when switching between local control with active panels and dispatcher control with panel actions disabled.

Would it be good to have an action that enables/disables a panel (PE, CPE, LE)? An action that enables/disables all Positionables on the panel and if it's LE, enables/disables all the LayoutTurnouts?

Would it be good to have an action RunOnce? An action that has a single child action that's only executed one time.

dsand47 commented:
> Would it be good to have an action that enables/disables a panel (PE, CPE, LE)? An action that enables/disables all Positionables on the panel and if it's LE, enables/disables all the LayoutTurnouts?

That capability is already built into the panel options.  

> Would it be good to have an action RunOnce? An action that has a single child action that's only executed one time.

Where would this be used as opposed to `TriggerOnce`?

danielb987 commented:
> > Would it be good to have an action that enables/disables a panel (PE, CPE, LE)? An action that enables/disables all Positionables on the panel and if it's LE, enables/disables all the LayoutTurnouts?
> 
> That capability is already built into the panel options.

LayoutEditor has the option `Allow Layout Control`. But I don't see that option in the other editors. And as far as I know, there is no way to control LEs `Allow Layout Control` from LogixNG. I thought about that when I read your comment above:

`This will be useful, especially with for each, when switching between local control with active panels and dispatcher control with panel actions disabled.`

.

> > Would it be good to have an action RunOnce? An action that has a single child action that's only executed one time.
> 
> Where would this be used as opposed to `TriggerOnce`?

It would be a bit more intuitive than to setup TriggerOnce. Instead of having
```
IfThenElse
  ? If TriggerOnce
    ? E Always true
  ! Then
```
you would have
```
RunOnce
  ! A
```
Since others have suggested using Sequence for this, I thought that it might be a good idea to have an action for it.

dsand47 commented:
> LayoutEditor has the option `Allow Layout Control`. But I don't see that option in the other editors. And as far as I know, there is no way to control LEs `Allow Layout Control` from LogixNG. I thought about that when I read your comment above:

For PE and CPE the option is `Panel items control layout`.

> > > Would it be good to have an action RunOnce? An action that has a single child action that's only executed one time.

After come consideration, I think `RunOnce` would be a good alternative to `TriggerOnce`.

danielb987 commented:
> > LayoutEditor has the option `Allow Layout Control`.
> 
> For PE and CPE the option is `Panel items control layout`.

Would it be useful to be able to control this from LogixNG? For example when switching between local control with active panels and dispatcher control with panel actions disabled?

KenC57 commented:
One trick I've used is to flag some things, like sensors, to only show when not
in edit mode. For testing I'll have things like occupancy sensors on the panel.
That's so I can test the panel offline. But it can crowd up a panel a lot too.
What I've not tried is clicking where something exists but hidden in the
non-edit mode.

danielb987 commented:
> After come consideration, I think `RunOnce` would be a good alternative to `TriggerOnce`.

The `RunOnce` action is added in PR #11650.
